File: 2024/07/7.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"strconv"
	"strings"
)

type Equation struct {
	result   int
	operands []int
}

func add(a, b int) int {
	return a + b
}
func mul(a, b int) int {
	return a * b
}
func concat(a, b int) int {
	res, _ := strconv.Atoi(strconv.Itoa(a) + strconv.Itoa(b))
	return res
}

func parseInput() []Equation {
	equations := []Equation{}
	scanner := bufio.NewScanner(os.Stdin)
	for scanner.Scan() {
		parts := strings.Split(scanner.Text(), ": ")
		result, _ := strconv.Atoi(parts[0])
		values := strings.Split(parts[1], " ")
		operands := make([]int, len(values))
		for i, v := range values {
			o, _ := strconv.Atoi(v)
			operands[i] = o
		}
		equations = append(equations, Equation{result, operands})
	}
	return equations
}

func findOperations(equation *Equation, i int, partial int, ops []func(a, b int) int) bool {
	if i == len(equation.operands) {
		return partial == equation.result
	} else {
		for _, op := range ops {
			if findOperations(equation, i+1, op(partial, equation.operands[i]), ops) {
				return true
			}
		}
		return false
	}
}

func main() {
	equations := parseInput()
	total1, total2 := 0, 0
	part1Ops := []func(a, b int) int{add, mul}
	part2Ops := []func(a, b int) int{add, mul, concat}

	for _, equation := range equations {
		if findOperations(&equation, 1, equation.operands[0], part1Ops) {
			total1 += equation.result
			total2 += equation.result
		} else if findOperations(&equation, 1, equation.operands[0], part2Ops) {
			total2 += equation.result
		}
	}

	fmt.Println(total1)
	fmt.Println(total2)
}

File: 2024/07/7.rb
#!/usr/bin/env ruby

equations = []
$stdin.readlines.each do |line|
  result, values = line.split(': ')
  values = values.split(' ').map(&:to_i)
  equations << [result.to_i, values]
end

module AoC
  refine Integer do
    def concat(other)
      (self.to_s + other.to_s).to_i
    end
  end
end
using AoC

# Depth-first search for equation rule, branching on each operation
def find_ops(result, values, ops, partial)
  return false if partial > result # optimization to prune dead trees
  if values.size == 1
    ops.any? do |op|
      result == partial.send(op, values[0])
    end
  else
    ops.any? do |op|
      find_ops(result, values[1..], ops, partial.send(op, values[0]))
    end
  end
end

total1, total2 = [0, 0]
equations.each do |result, values|
  if find_ops(result, values[1..], %i[+ *], values[0])
    total1 += result
    total2 += result
  elsif find_ops(result, values[1..], %i[+ * concat], values[0])
    total2 += result
  end
end
puts total1
puts total2
